Fix data races

```
go test -race .
Check &{next:0 past:[9223372036854775807 9223372036854775807 9223372036854775807] lastFinished:0 curr:10} --- &{next:0 past:[0 0 0] st:2 cleanSlate:0 rt:0xc42007ef68} -- &{gcNext:0 gcEstimation:0 gcStart:{sec:0 nsec:0 loc:<nil>} gcPast:[0 0 0] reqCount:0 reqPast:[0 0 0] reqMean:0 reqVar:0 reqEstimation:0}

Check &{next:0 past:[9223372036854775807 9223372036854775807 9223372036854775807] lastFinished:0 curr:10} --- &{next:0 past:[0 0 0] st:2 cleanSlate:0 rt:0xc42007ef68} -- &{gcNext:0 gcEstimation:0 gcStart:{sec:0 nsec:0 loc:<nil>} gcPast:[0 0 0] reqCount:10 reqPast:[0 0 0] reqMean:16652.800000000003 reqVar:2.3280602805600002e+10 reqEstimation:0}

Begin &{next:0 past:[9223372036854775807 9223372036854775807 9223372036854775807] lastFinished:0 curr:10} --- &{next:0 past:[0 0 0] st:2 cleanSlate:0 rt:0xc42007ef68} -- &{gcNext:0 gcEstimation:0 gcStart:{sec:0 nsec:0 loc:<nil>} gcPast:[0 0 0] reqCount:10 reqPast:[0 0 0] reqMean:16652.800000000003 reqVar:2.3280602805600002e+10 reqEstimation:0}

==================
WARNING: DATA RACE
Write at 0x00c420096988 by goroutine 11:
  github.com/gcinterceptor/gci-go/gccontrol.(*unavailabilityEstimator).gcFinished()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/unavailability.go:30 +0xf2
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before.func1()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:88 +0x2f2

Previous read at 0x00c420096988 by goroutine 10:
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).shed()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:108 +0xfc
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:98 +0x399
  github.com/gcinterceptor/gci-go/gccontrol.TestInterceptor()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor_test.go:33 +0x7fa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107

Goroutine 11 (running) created at:
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:97 +0x388
  github.com/gcinterceptor/gci-go/gccontrol.TestInterceptor()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor_test.go:33 +0x7fa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107

Goroutine 10 (running) created at:
  testing.(*T).Run()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:882 +0xaa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107
  testing.runTests()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:888 +0x4e0
  testing.(*M).Run()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:822 +0x1c3
  main.main()
      github.com/gcinterceptor/gci-go/gccontrol/_test/_testmain.go:50 +0x20f
==================
==================
WARNING: DATA RACE
Write at 0x00c4200969f0 by goroutine 11:
  github.com/gcinterceptor/gci-go/gccontrol.(*unavailabilityEstimator).gcFinished()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/unavailability.go:40 +0x23a
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before.func1()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:88 +0x2f2

Previous read at 0x00c4200969f0 by goroutine 10:
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).shed()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:108 +0xc9
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:98 +0x399
  github.com/gcinterceptor/gci-go/gccontrol.TestInterceptor()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor_test.go:33 +0x7fa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107

Goroutine 11 (running) created at:
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:97 +0x388
  github.com/gcinterceptor/gci-go/gccontrol.TestInterceptor()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor_test.go:33 +0x7fa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107

Goroutine 10 (running) created at:
  testing.(*T).Run()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:882 +0xaa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107
  testing.runTests()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:888 +0x4e0
  testing.(*M).Run()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:822 +0x1c3
  main.main()
      github.com/gcinterceptor/gci-go/gccontrol/_test/_testmain.go:50 +0x20f
==================
==================
WARNING: DATA RACE
Write at 0x00c420084970 by goroutine 11:
  sync/atomic.StoreInt32()
      /home/fireman/.gvm/gos/go1.8/src/runtime/race_amd64.s:229 +0xb
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before.func1()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:95 +0x340

Previous read at 0x00c420084970 by goroutine 10:
  github.com/gcinterceptor/gci-go/gccontrol.TestInterceptor()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor_test.go:43 +0x9a2
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107

Goroutine 11 (running) created at:
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:97 +0x388
  github.com/gcinterceptor/gci-go/gccontrol.TestInterceptor()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor_test.go:33 +0x7fa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107

Goroutine 10 (running) created at:
  testing.(*T).Run()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:882 +0xaa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107
  testing.runTests()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:888 +0x4e0
  testing.(*M).Run()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:822 +0x1c3
  main.main()
      github.com/gcinterceptor/gci-go/gccontrol/_test/_testmain.go:50 +0x20f
==================
End &{next:0 past:[9223372036854775807 9223372036854775807 9223372036854775807] lastFinished:10 curr:10} -- &{next:1 past:[3 0 0] st:3 cleanSlate:0 rt:0xc42007ef68} -- &{gcNext:1 gcEstimation:3458 gcStart:{sec:0 nsec:0 loc:<nil>} gcPast:[3458 0 0] reqCount:0 reqPast:[67512 0 0] reqMean:0 reqVar:0 reqEstimation:67512}

==================
WARNING: DATA RACE
Read at 0x00c420084960 by goroutine 10:
  runtime.convT2E()
      /home/fireman/.gvm/gos/go1.8/src/runtime/iface.go:191 +0x0
  github.com/gcinterceptor/gci-go/gccontrol.TestInterceptor()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor_test.go:49 +0xa1a
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107

Previous write at 0x00c420084960 by goroutine 11:
  sync/atomic.StoreInt64()
      /home/fireman/.gvm/gos/go1.8/src/runtime/race_amd64.s:234 +0xb
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before.func1()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:91 +0x30c

Goroutine 10 (running) created at:
  testing.(*T).Run()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:882 +0xaa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107
  testing.runTests()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:888 +0x4e0
  testing.(*M).Run()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:822 +0x1c3
  main.main()
      github.com/gcinterceptor/gci-go/gccontrol/_test/_testmain.go:50 +0x20f

Goroutine 11 (finished) created at:
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:97 +0x388
  github.com/gcinterceptor/gci-go/gccontrol.TestInterceptor()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor_test.go:33 +0x7fa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107
==================
==================
WARNING: DATA RACE
Read at 0x00c420084968 by goroutine 10:
  runtime.convT2E()
      /home/fireman/.gvm/gos/go1.8/src/runtime/iface.go:191 +0x0
  github.com/gcinterceptor/gci-go/gccontrol.TestInterceptor()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor_test.go:50 +0xab3
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107

Previous write at 0x00c420084968 by goroutine 11:
  sync/atomic.StoreInt64()
      /home/fireman/.gvm/gos/go1.8/src/runtime/race_amd64.s:234 +0xb
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before.func1()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:92 +0x323

Goroutine 10 (running) created at:
  testing.(*T).Run()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:882 +0xaa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107
  testing.runTests()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:888 +0x4e0
  testing.(*M).Run()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:822 +0x1c3
  main.main()
      github.com/gcinterceptor/gci-go/gccontrol/_test/_testmain.go:50 +0x20f

Goroutine 11 (finished) created at:
  github.com/gcinterceptor/gci-go/gccontrol.(*Interceptor).Before()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor.go:97 +0x388
  github.com/gcinterceptor/gci-go/gccontrol.TestInterceptor()
      /home/fireman/Workspace/go/src/github.com/gcinterceptor/gci-go/gccontrol/interceptor_test.go:33 +0x7fa
  testing.tRunner()
      /home/fireman/.gvm/gos/go1.8/src/testing/testing.go:657 +0x107
==================
--- FAIL: TestInterceptor (0.01s)
        testing.go:610: race detected during execution of test
FAIL
FAIL    github.com/gcinterceptor/gci-go/gccontrol       0.018
```

--- a/gccontrol/interceptor_test.go
+++ b/gccontrol/interceptor_test.go
@@ -1,6 +1,7 @@
 package gccontrol
 
 import (
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -40,7 +41,7 @@ func TestInterceptor(t *testing.T) {
 	i.After(sr1)
 
 	for {
-		if i.doingGC == 0 {
+		if atomic.LoadInt32(&i.doingGC) == 0 {
 			break
 		}
 		time.Sleep(10 * time.Millisecond)

--- a/gccontrol/unavailability.go
+++ b/gccontrol/unavailability.go
@@ -2,32 +2,33 @@ package gccontrol
 
 import (
 	"math"
+	"sync/atomic"
 	"time"
 )
 
 func newUnavailabilityEstimator(size int) *unavailabilityEstimator {
 	return &unavailabilityEstimator{
-		gcPast:  make([]time.Duration, size),
-		reqPast: make([]time.Duration, size),
+		gcPast:  make([]int64, size),
+		reqPast: make([]int64, size),
 	}
 }
 
 type unavailabilityEstimator struct {
-	gcNext       int             // Next index in gcPast.
-	gcEstimation time.Duration   // Current estimation of the next GC duration.
-	gcStart      time.Time       // Last GC start time.
-	gcPast       []time.Duration // History of garbage collection duration estimations.
+	gcNext       int       // Next index in gcPast.
+	gcEstimation int64     // Current estimation of the next GC duration in nanoseconds.
+	gcStart      time.Time // Last GC start time.
+	gcPast       []int64   // History of garbage collection duration estimations in nanoseconds.
 
-	reqCount        int             // Number of requests finished since last collection.
-	reqPast         []time.Duration // History of request duration estimations.
-	reqMean, reqVar float64         // Request statistics in nanoseconds. Made them float to make math easier.
-	reqEstimation   time.Duration   // Current request duration estimation.
+	reqCount        int     // Number of requests finished since last collection.
+	reqPast         []int64 // History of request duration estimations in nanoseconds.
+	reqMean, reqVar float64 // Request statistics in nanoseconds. Made them float to make math easier.
+	reqEstimation   int64   // Current request duration estimation in nanoseconds.
 }
 
 func (u *unavailabilityEstimator) gcFinished(d time.Duration) {
 	// Estimate GC duration.
-	u.gcPast[u.gcNext] = d
-	u.gcEstimation = maxDuration(u.gcPast)
+	u.gcPast[u.gcNext] = d.Nanoseconds()
+	atomic.StoreInt64(&u.gcEstimation, maxDuration(u.gcPast))
 
 	// Estimate the time processing a request.
 	// Using 68–95–99.7 rule to have a good coverage on the request size.
@@ -36,8 +37,9 @@ func (u *unavailabilityEstimator) gcFinished(d time.Duration) {
 	if u.reqCount > 1 {
 		stdDev = math.Sqrt(u.reqVar) / float64(u.reqCount-1)
 	}
-	u.reqPast[u.gcNext] = time.Duration(u.reqMean+(3*stdDev)) * time.Nanosecond
-	u.reqEstimation = maxDuration(u.reqPast)
+	u.reqPast[u.gcNext] = int64(u.reqMean + (3 * stdDev))
+	atomic.StoreInt64(&u.reqEstimation, maxDuration(u.reqPast))
+
 	u.reqCount = 0
 	u.reqMean = 0
 	u.reqVar = 0
@@ -45,11 +47,11 @@ func (u *unavailabilityEstimator) gcFinished(d time.Duration) {
 }
 
 func (u *unavailabilityEstimator) estimate(queueSize int64) time.Duration {
-	trailingReqs := time.Duration(queueSize * u.reqEstimation.Nanoseconds())
+	trailingReqs := queueSize * atomic.LoadInt64(&u.reqEstimation)
 	if trailingReqs < 0 {
 		trailingReqs = 0
 	}
-	return u.gcEstimation + trailingReqs
+	return time.Duration(atomic.LoadInt64(&u.gcEstimation) + trailingReqs)
 }
 
 // Flags that a request has been finished. This is needed to estimate the request duration. The latter
@@ -70,8 +72,8 @@ func (u *unavailabilityEstimator) requestFinished(d time.Duration) {
 	}
 }
 
-func maxDuration(s []time.Duration) time.Duration {
-	max := 0 * time.Nanosecond
+func maxDuration(s []int64) int64 {
+	max := int64(0)
 	for _, d := range s {
 		if d > max {
 			max = d

